doc/design/csp.md

```diff
@@ -89,6 +89,55 @@ The point here is that we need a consistent way to compose types, like in C++ we
 
 ### Select
 
+In Go, the `select` statement lets a goroutine wait on multiple communication operations. A `select` blocks untill one of its cases can run, then it executes that case. It chooses one at random if multiple are ready.
+
+```go
+
+ch1  := make(chan int)       
+ch2  := make(chan int, 100)
+
+x := 0
+
+for {
+    select {
+    case ch1 <- x:
+      x := x + 1
+    case y <- ch2:
+      fmt.Println("Received on channel")
+    default:
+      fmt.Println("Default")
+    }
+  }
+
+```
+
+In Fluid, we should be able to do the same:
+
+```python
+ch1  = fluid.make_chan(dtype=INT)
+ch2 = fluid.make_chan(dtype=INT, 100)
+
+select = fluid.select()
+
+with sel.case(ch1, 'w', X):
+    fluid.layers.increment(X)
+
+with sel.case(ch2, 'r', Y):
+    fluid.print("Received on Channel")
+
+with sel.default():
+    fluid.print("Default")
+
+```
+
+In the above code snippet, `X` and `Y` are variables. Now let us look at each of these statements one by one.
+
+- `sel.case(ch1, 'w', X)` : This specifies that we are writing to `ch1` and we want to write the integer in variable `X` to the channel. The character `w` is used here to make the syntax familar to write syntax in Python I/O.
+
+- `sel.case(ch2, 'r', Y)` : This specifies that we would like to read the result from `ch2` into variable `Y`. The character `r` is used here to make the syntax familar to read syntax in Python I/O.
```

Comment:
An alternative to consider: `sel.case(ch1.recv(x))` and `sel.case(ch2.send(Y))`.

Reply:
My proposal is similar -- Defining the `send` and `recv` (write/read) for the Channel as @helinwang proposed above, or writing out `send` and `recv` as separate methods:

```
x = sel.case(fluid.recv(ch1))
sel.case(fluid.send(ch2, Y))```

The methods can have the definition like:

```
void send(Channel* channel, Element* val)

Element* Receive(Channel* channel)
```

Reply:
I agree the C++ syntax should look like the one that you mentioned above. I am not sure how we would be able to implement `sel.case(ch1.recv(x))` in python. This is because `ch1.recv(x)` can also be called in a standalone fashion, outside of `select`. I think this might depend on how we design the `select` operator in Paddle.

Reply:
Sorry for the late reply!

I have not find `recv` design in this doc, but I think `ch1.recv(x)` should return a var, so could work both in standalone and in select.

Reply:
I agree with @helinwang 's proposal as well.

Reply:
I agree with @helinwang as well. I am approving this PR so we could merge it, together with @kavyasrinet 's PR on send/recv syntax. Then, let us rethink about them.